### PR TITLE
passing batch outputs to on_train_batch_end

### LINKED ISSUE #4326: Accessing performance of model in progress bar
When I run training, I see progress bar indicate iterations/sec. How can I access it? I wrote a simple hook:

```
class PerfCallback(ProgressBar):

    def __init__(self):
        super().__init__()  # don't forget this :)

    def on_train_start(self, trainer, pl_module):
        super().on_train_start(trainer, pl_module)
        self.total_runtime = 0
        self.unpadded_tokens = 0
        self.all_tokens = 0
        self.total_steps = 0

    def on_train_batch_start(self, trainer, pl_module, batch, batch_idx, dataloader_idx):
        self.t0 = time.time()

    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
        runtime = time.time() - self.t0
        self.total_runtime += runtime
        print("on batch end runtime=%.2f, it/s = %.2f" %(runtime, 1/runtime))
```

However, my prints indicate ~1.7it/s but the progress bar sows 6.07s/it. 

In my training_step, I also return some stats (batch size, number of tokens) in logs by returning the following:
```
{"loss": loss_tensors[0], "log":logs, "progress_bar":{"global_step":self.global_step}}
```
 for more relevant perf metrics. However, in my callback. print(outputs) shows an empty list. Anything I'm missing?

### PR DESCRIPTION
**Pull request by swethmandava**
## What does this PR do?

Passes batch outputs to on_train_batch_end instead of epoch end outputs

Fixes #4326 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [Y] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by edenlightning** [comment]
Thanks for the PR!

Please take a second look as it seems test are failing...

**Comment by Borda** [comment]
@swethmandava could you please check failing tests and also add a test for this change?

**Comment by swethmandava** [comment]
@Borda @edenlightning fixed pytest errors. it now fails with `/bin/bash: /root/miniconda3/envs/lightning/lib/libtinfo.so.6: no version information available (required by /bin/bash)` seems to be conda version related - can you help?

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -268,12 +268,26 @@ def reset_train_val_dataloaders(self, model):
...
         # track the outputs to reduce at the end of the epoch
-        for opt_idx, opt_outputs in enumerate(epoch_end_outputs):
+        for opt_idx, opt_outputs in enumerate(batch_end_outputs):
+            sample_output = opt_outputs[-1]
+
+            # decide if we need to reduce at the end of the epoch automatically
+            auto_reduce_tng_result = isinstance(sample_output, Result) and sample_output.should_reduce_on_epoch_end
+            hook_overridden = (
+                is_overridden("training_epoch_end", model=self.trainer.get_model()) or
+                is_overridden("on_train_epoch_end", model=self.trainer.get_model()))
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:281]:
```suggestion
                is_overridden("on_train_epoch_end", model=self.trainer.get_model())
            )
```

**Review by Borda — APPROVED** [review]
feel like this name is cleaner, but it may hear some users who pass args as kwargs...
cc: @tchaton @SeanNaren

**Comment by swethmandava** [comment]
Not sure what changed. `is_overridden("on_train_epoch_end", model=self.trainer.get_model())` returns False even when overwritten in callback @SeanNaren . Hence will have to implement `training_epoch_end`

**Comment by awaelchli** [comment]
@swethmandava `is_overridden` refers to methods in LightningModule. It has nothing to do with methods implemented in the Callback class

**Comment by swethmandava** [comment]
> @swethmandava `is_overridden` refers to methods in LightningModule. It has nothing to do with methods implemented in the Callback class

Since epoch_end_outputs are passed to callbacks, it would make sense to also check if specific hooks like `on_train_epoch_end` has been implemented in the callbacks to decide if we should track outputs. Any good way to do it? @awaelchli

**Comment by awaelchli** [comment]
The Callback class has these methods so they are "implemented" just not necessarily overridden but that doesn't matter, they are no-op so you can just pass in the outputs right?
I see no need to add new logic.

**Comment by swethmandava** [comment]
> The Callback class has these methods so they are "implemented" just not necessarily overridden but that doesn't matter, they are no-op so you can just pass in the outputs right?
> I see no need to add new logic.

They are only tracked if needed https://github.com/PyTorchLightning/pytorch-lightning/blob/056602001c72283c2ffdd028941f89176666f191/pytorch_lightning/trainer/training_loop.py#L280

so unless `training_epoch_end` is overridden, `on_train_epoch_end` receives an empty list of outputs. Do you want to track 'always' instead?

**Comment by BotScutters** [comment]
Hey all, I'm working on a callback that would generate a plot of model outputs each epoch and encountering an issue that appears to be related to this. I'd expect I should be able to grab the model outputs from the `outputs` arg of `on_training_batch_end`, `on_validation_batch_end`, etc., but this shows comes in as `None`. A bit of googling brought me here, and it seems this issue has gone a little stale.

Is this issue still being worked on? If not, is there a suggested workaround? Thanks!

**Comment by swethmandava** [comment]
@SeanNaren can this be merged?

**Review by s-rog — APPROVED** [review]

**Review by tchaton — COMMENTED** [review]
Would it be possible to have a test ?

**Comment by SeanNaren** [comment]
Apologies for the stale! A simple boring model test that ensures that if the user overrides this hook, they retrieve the appropriate model outputs would be nice :)

Similar to this: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/models/test_hooks.py#L52

thanks for working on this!

**Comment by swethmandava** [comment]
> Apologies for the stale! A simple boring model test that ensures that if the user overrides this hook, they retrieve the appropriate model outputs would be nice :)
> 
> Similar to this: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/models/test_hooks.py#L52
> 
> thanks for working on this!

Done!

**Review by SeanNaren — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by SeanNaren** [comment]
Thanks @swethmandava for doing this and apologies on the delay, fell through the cracks a few times...

File: tests/models/test_hooks.py [hunk 2]
```diff
@@ -90,6 +92,58 @@ def training_epoch_end(self, outputs):
         assert metrics[f'epoch_metric_{i}'] == i
 
 
+def test_training_epoch_end_metrics_collection_on_override(tmpdir):
+    """ Test that batch end metrics are collected when training_epoch_end is overridden at the end of an epoch. """
+    num_epochs = 1
+
+    class LoggingCallback(pl.Callback):
+
+        def on_train_epoch_end(self, trainer, pl_module):
```
Comment (Borda) [inline tests/models/test_hooks.py:101]: @SeanNaren flake8 found this bug, but it was still merged, could pls fix it in #5666

**Pull request merged**